agent/testflinger_agent/agent.py

```diff
@@ -179,168 +116,71 @@ def mark_device_offline(self):
-        job_id = job_data["job_id"]
-
-        with tempfile.NamedTemporaryFile(suffix="tar.gz") as archive_tmp:
-            archive_path = Path(archive_tmp.name)
-            # download attachment archive
-            logger.info(f"Downloading attachments for {job_id}")
-            self.client.get_attachments(job_id, path=archive_path)
-            # extract archive into the attachments folder
-            logger.info(f"Unpacking attachments for {job_id}")
-            with tarfile.open(archive_path, "r:gz") as tar:
-                tar.extractall(cwd / ATTACHMENTS_DIR, filter=secure_filter)
-
-        # side effect: remove all attachment data from `job_data`
-        # (so there is no interference with existing processes, especially
-        # provisioning or firmware update, which are triggered when these
-        # sections are not empty)
-        for phase in (
-            TestPhase.PROVISION,
-            TestPhase.FIRMWARE_UPDATE,
-            TestPhase.TEST,
-        ):
-            phase_str = f"{phase}_data"
-            try:
-                phase_data = job_data[phase_str]
-            except KeyError:
-                pass
-            else:
-                # delete attachments, if they exist
-                phase_data.pop("attachments", None)
-                # it may be the case that attachments were the only data
-                # included for this phase, so the phase can now be removed
-                if not phase_data:
-                    del job_data[phase_str]
-
     def process_jobs(self):
         """Coordinate checking for new jobs and handling them if they exists"""
 
-        TEST_PHASES = [
-            TestPhase.SETUP,
-            TestPhase.PROVISION,
-            TestPhase.FIRMWARE_UPDATE,
-            TestPhase.TEST,
-            TestPhase.ALLOCATE,
-            TestPhase.RESERVE,
-        ]
-
         # First, see if we have any old results that we couldn't send last time
         self.retry_old_results()
 
         self.check_restart()
         job_data = self.client.check_jobs()
         while job_data:
+            job_id = job_data["job_id"]
             try:
+                # Create the job
                 job = TestflingerJob(job_data, self.client)
-                event_emitter = EventEmitter(
-                    job_data.get("job_queue"),
-                    job_data.get("job_status_webhook"),
-                    self.client,
-                    job.job_id,
-                )
-                job_end_reason = TestEvent.NORMAL_EXIT
-
-                logger.info("Starting job %s", job.job_id)
-                event_emitter.emit_event(
-                    TestEvent.JOB_START,
-                    f"{self.client.server}/jobs/{job.job_id}",
-                )
-                rundir = os.path.join(
-                    self.client.config.get("execution_basedir"), job.job_id
-                )
-                os.makedirs(rundir)
```

Comment:
Not really done testing this, but wanted to get started on it... I did see this so far. Because it never forces creation of this dir, if it's missing, then it'll result in problems later:
```
[24-10-08 19:01:02]   ERROR: (cmd.py:32)| cannot access local variable 'job' where it is not associated with a value
Traceback (most recent call last):
  File "/home/plars/work/cert/testflinger/agent/testflinger_agent/cmd.py", line 27, in main
    start_agent()
  File "/home/plars/work/cert/testflinger/agent/testflinger_agent/__init__.py", line 139, in start_agent
    agent.process_jobs()
  File "/home/plars/work/cert/testflinger/agent/testflinger_agent/agent.py", line 159, in process_jobs
    if job.go(phase):
       ^^^
UnboundLocalError: cannot access local variable 'job' where it is not associated with a value
```

Reply:
> Because it never forces creation of this dir, if it's missing

The rundir is created [here](https://github.com/canonical/testflinger/blob/41f8a6d2281eda3e505f803f3768cbd024c5ee5b/agent/testflinger_agent/job.py#L505), when the `job` object is `__init__`ed.

That error is raised in the `finally` clause because the `job` object was never properly created so there should be an indication in the agent log about what the actual error was. I just don't think it's the `rundir` not being created.

Reply:
Alright, I think I figured this out. `Path.mkdir` needs the `parents` argument set to `True` in order to mirror how `os.mkdirs` creates intermediate directories recursively. This should be fixed now